ExtraDry/ExtraDry.Server/RuleEngine.cs

```diff
@@ -185,42 +146,195 @@ public UndeleteResult Undelete<T>(T item)
-        if(AttemptSoftDelete(item)) {
-            await commit();
-            result = DeleteResult.SoftDeleted;
-        }
+
         try {
             await remove();
             await commit();
             result = DeleteResult.HardDeleted;
         }
-        catch { // Hmmm, can we be specific about exception type here?
-            if(result == DeleteResult.NotDeleted) {
-                throw new InvalidOperationException("Item could not be hard-deleted and no rules for soft-delete provided.");
+        catch {
+            //Do not throw exceptions on commit- Just return appropriate result.
+        }
+        return result;
+    }
+
+    /// <summary>
+    /// Register the remove action to be used when hard-deleting an item.
+    /// </summary>
+    /// <param name="remove">The action that is executed for a hard-delete.</param>
+    public void RegisterRemove<T>(Action<T> remove)
+    {
+        RemoveFunctors.Add(typeof(T), e => remove((T)e));
+    }
+
+    /// <summary>
+    /// Register the commit action to be used when hard-deleting an item.
+    /// </summary>
+    /// <param name="commit">The action that is executed to commit hard-delete.</param>
+    public void RegisterCommit(Func<Task> commit)
+    {
+        CommitFunctor = commit;
+    }
+
+    /// <inheritdoc cref="TryHardDeleteAsync(object[])" />
+    public async Task<DeleteResult> TryHardDeleteAsync<T>(T item) where T : class
+    {
+        Action<object> remove;
+        try {
+            remove = RemoveFunctors[typeof(T)];
+        }
+        catch(KeyNotFoundException ex) {
+            throw new DryException($"Could not find registered expression for removing Entity. Register Entity removal with the RegisterRemove method.", ex);
+        }
+
+        var result = DeleteResult.NotDeleted;
+
+        try {
+            remove(item);
+            await CommitFunctor();
+            result = DeleteResult.HardDeleted;
+        }
+        catch {
+            //Do not throw exceptions on commit - Just return appropriate result.
+        }
+        RemoveFunctors.Clear();
+        return result;
+    }
+
+    /// <summary>
+    /// Processes a Hard Delete for multiple items.
+    /// Uses the remove actions from <see cref="RegisterRemove{T}(Action{T})"/> and commit action from <see cref="RegisterCommit(Func{Task})"/>.
+    /// </summary>
+    /// <param name="items">Items to delete</param>
+    public async Task<DeleteResult> TryHardDeleteAsync(params object[] items)
+    {
+        var result = DeleteResult.NotDeleted;
+
+        //Check all RemoveFunctors are available
+        try {
+            foreach(var item in items) {
+                Action<object>? remove;
+
+                //If item is an IEnumerable check its RemoveFunctor
+                var enumerableItem = item as IEnumerable;
+                if(enumerableItem != null) {
+                    remove = RemoveFunctors[enumerableItem.GetEnumerator().Current.GetType()];
```

Comment:
I can see that this works and it's OK, however it does allocate an enumerator that is then discarded. I think you can get this with `enumerableItem.GetType().GetGenericArguments().First()` more efficiently.

Reply:
Fixed